File: app/javascripts/models/line.js
//= require "../model"

var Line = new JS.Class(Model, {
  extend: {
    resource: 'line',
    columns: ['id', 'transaction_id', 'item_id', 'title', 'description', 'quantity', 'condition', 'discount', 'price', 'credit', 'cash', 'purchase', 'taxable', 'discountable'],
    belongs_to: ['item', 'transaction'],
    'title': {
      'presence_of': {}
    }
  },
  
  subtotal: function() {
    if(this.purchase) {
      return Math.round(this.quantity * this.discount * this.price);
    } else {
      return Math.round(this.quantity * this.condition * this.credit * -1);
    }
  },
  
  creditSubtotal: function() {
    if(!this.purchase) {
      return parseInt(Math.round(this.quantity * this.credit * this.condition * -1));
    } else {
      return 0;
    }
  },
  
  cashSubtotal: function() {
    if(!this.purchase) {
      return parseInt(Math.round(this.quantity * this.cash * this.condition * -1));
    } else {
      return 0;
    }
  }
});

File: app/models/line.rb
class Line < ActiveRecord::Base
  validates_presence_of         :title, :quantity, :condition, :discount, :price, :credit, :cash
  validates_inclusion_of        :purchase, :in => [true, false]
  validates_inclusion_of        :taxable, :in => [true, false]
  
  after_initialize              :_default
  
  belongs_to                    :account
  belongs_to                    :transaction
  belongs_to                    :item
  belongs_to                    :unit
  
  # Calculate subtotal in pennies
  def subtotal
    if self.purchase?
      (self.quantity * self.discount * self.condition * self.price).round.to_i
    else
      (self.quantity * self.discount * self.condition * self.credit * -1).round.to_i
    end
  end
  
  # Is line purchase?
  def purchase?
    self.purchase
  end
  
  # Is line taxable?
  def taxable?
    self.taxable
  end
  
  private
  
  def _default
    self.quantity   ||= 1 if new_record?
    self.condition  ||= 1 if new_record?
    self.discount   ||= 1 if new_record?
    self.price      ||= 0 if new_record?
    self.credit     ||= 0 if new_record?
    self.cash       ||= 0 if new_record?
    self.purchase   = true if new_record? && self.purchase.nil?
    self.taxable    = true if new_record? && self.taxable.nil?
  end
end